src/lib.rs

```diff
@@ -384,21 +384,50 @@ impl Drop for Context {
 
 pub struct Socket {
     sock: *mut c_void,
-    closed: bool
+    owned: bool,
 }
 
 unsafe impl Send for Socket {}
 
 impl Drop for Socket {
     fn drop(&mut self) {
-        match self.close_final() {
-            Ok(()) => { debug!("socket dropped") },
-            Err(e) => panic!(e)
+        if self.owned {
+            if unsafe { zmq_sys::zmq_close(self.sock) } == -1i32 {
+                panic!(errno_to_error());
+            } else {
+                debug!("socket dropped");
+            }
         }
     }
 }
 
 impl Socket {
+    /// Consume the Socket and return the raw socket pointer.
+    ///
+    /// Failure to close the raw socket manually or call `from_raw`
+    /// will lead to a memory leak.
+    pub fn into_raw(mut self) -> *mut c_void {
```

Comment:
Thinking about it more - as we are comparing the `Box` ptr conventions with `slice`, we should take into account that a `Box` is an owned type, whereas a `slice` isn't. Semantically it makes sense that a `Box` can provide an owned pointer like `into_raw` that can outlive the box, whereas a `slice` is a reference so you can't apply the same logic. I.e. you can't call `into_raw(self)` on a slice because you can't get an owned slice.

I know that's a flimsy argument, but as a noob it seems that `Socket` is more akin to `Box` than a `slice`. What do you think?